src/storage/database.rs

```diff
@@ -21,24 +21,56 @@ impl DatabaseBackend {
         Ok(conn.query(query, &[&path])?[0].get(0))
     }
 
-    pub(super) fn get(&self, path: &str, max_size: usize) -> Result<Blob, Error> {
+    pub(super) fn get(
+        &self,
+        path: &str,
+        max_size: usize,
+        range: Option<FileRange>,
+    ) -> Result<Blob, Error> {
         use std::convert::TryInto;
 
         // The maximum size for a BYTEA (the type used for `content`) is 1GB, so this cast is safe:
         // https://www.postgresql.org/message-id/162867790712200946i7ba8eb92v908ac595c0c35aee%40mail.gmail.com
         let max_size = max_size.min(std::i32::MAX as usize) as i32;
 
-        // The size limit is checked at the database level, to avoid receiving data altogether if
-        // the limit is exceeded.
-        let rows = self.pool.get()?.query(
-            "SELECT
-                 path, mime, date_updated, compression,
-                 (CASE WHEN LENGTH(content) <= $2 THEN content ELSE NULL END) AS content,
-                 (LENGTH(content) > $2) AS is_too_big
-             FROM files
-             WHERE path = $1;",
-            &[&path, &(max_size)],
-        )?;
+        let rows = if let Some(r) = range {
+            // when we only want to get a range we can validate already if the range is small enough
+            let (range_size, _) = r.size_hint();
+            if range_size > max_size as usize {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    crate::error::SizeLimitReached,
+                )
+                .into());
+            }
+            let range_start = i32::try_from(*r.start())?;
```

Comment:
is there another way to ensure this? since `.len` doesn't exist on that range. 

We also could switch to another type for the ranges, or only 32 bit integers, for that matter.

Reply:
Hmm, the difference only matters if the archive is over 4 GB, right? Even stm32f4 is only 140 MB, I think 4 GB should be ok. That's the size at which we should probably start imposing size limits anyway.

Reply:
I changed the range to be wrapped in a newtype to have a nice `.len` and not to have potential undefined behavior.